control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go

```diff
@@ -115,6 +115,7 @@ func ReconcileRole(role *rbacv1.Role, ownerRef config.OwnerRef) error {
 			Resources: []string{
 				"secrets",
 				"services",
+				"nodes",
```

Comment:
this is given access to management cluster Nodes, can we please drop it?

Reply:
I added this because without it, the controller says it has no perms to list or watch node objects, and so the drain controller can't even run.

I probably am misunderstanding something though. Is there a way I would be able to fix that error?

Reply:
As I understand it the upgrader is really watching machinesets and deriving the node object https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/setup.go#L28-L34

Would I have to do the same for this?

Reply:
>As I understand it the upgrader is really watching machinesets and deriving the node object https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/setup.go#L28-L34
Would I have to do the same for this?

As per the link above, the upgrader is watching nodes in the guest cluster and machineSets in the management cluster. In this case we only want to watch Nodes.

>I added this because without it, the controller says it has no perms to list or watch node objects, and so the drain controller can't even run.

That should not be the case. The guest cluster watcher is built against the --target-kubeconfig. the rbac you are changing is for the in-cluster client i.e the management cluster.

Reply:
> That should not be the case. The guest cluster watcher is built against the manager client which uses the --target-kubeconfig. The rbac you are changing is for the in-cluster client i.e the management cluster.

Hmm let me remove this and test again. I think I might be doing something wrong, because if the inplaceupgrade controller doesn't complain, this really shouldn't either.

> The guest cluster watcher is built against the manager client which uses the --target-kubeconfig.

I do see this but I am not 100% certain I understand how this is being wired up. The controller I set up is just passing in `opts.Manager` which then seems to set up the client via the hosted cluster's client and not the targetconfig? Probably a dumb question and I am missing a connection piece somewhere

Reply:
The controller doesn't set up any watches during the `controller.New` call. What happens is that we


* Watch nodes without any specific cache config, which falls through to use the `Cache` from the `manager` that points at the hosted cluster by means if saving it in the `Controller` and then injecting it into the `Source`: https://github.com/openshift/hypershift/blob/9d587470832a034c0579141f00cee1e7ac5484b0/control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/setup.go#L28
* Watch MachineSets by explicitly setting a cache which is the one that points to the mgmt cluster: https://github.com/openshift/hypershift/blob/9d587470832a034c0579141f00cee1e7ac5484b0/control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/setup.go#L32 (note the `NewKindWithCache` vs the previous `&source.Kind{}`)

Reply:
and yeah, what alberto said is correct, this just sets up rbac for the mgmt cluster which we don't need (and if we do, there is a bug somewhere :) )

Reply:
Ah I see, thanks for the explanation! Testing without the perms now

Reply:
It seems to be working fine without it, so removed. I think maybe my original testing had another issue and thought this was the fix (but was actually something else). Thanks for the pointers!